doc/designs/hsm.md

```diff
@@ -0,0 +1,277 @@
+
+
+## How to Use
+
+Using an HSM should be largely invisible to users and administrators beyond passing additional options during installation. The options required and any pre-installion work are HSM-specific.
+
+It will not be possible to mix and match by default. PKI supports specifying the token values so a user can override these using --pki-config-override but it is, and will be, untested.
+
+### Installation
+
+#### CA
+
+The token name, module name and shared library must be provided to the
+CA and KRA installers. These define where the keys will be stored
+and the naming convention for them.
+
+The CA signing, OCSP, Audit and Subsystem private keys and certificates
+are generated and stored in the HSM.
+
+| Option | Description                                       |
+|:------------------------ | :------------------------------ |
+| --token-name             | NSS name for the token          |
+| --library-path           | Path to PKCS#11 shared library  |
+| --token-password         | Password for the token          |
+
+This information will be stored in new schema so that replicas can auto-detect when an HSM is configured.
+
+ipa-ca-install will accept the same options.
+
+```
+attributeTypes: (
+  2.16.840.1.113730.3.8.21.1.TBD
+  NAME 'ipaCaHSMConfiguration'
+  DESC 'HSM Configuration'
+  SYNTAX 1.3.6.1.4.1.1466.115.121.1.15
+  SINGLE-VALUE
+  X-ORIGIN 'IPA v4.10'
+)
+```
+This will be added as a MAY to the ```ipaCa``` objectclass.
+
+This attribute will be semi-colon delimited and contain the HSM information needed to install a replica in the form of:
+
+token-name;library-path
+
+The token password will be prompted by ipa-replica-install or passed on the cli.
+
+The presence of this attribute is enough to indicate that an HSM is present. If components are missing then the user must provide them on the cli. When ever a replica, replica CA, KRA or replica KRA is added this attribute will be examined to determine whether an HSM is available or not, and what the options are. The options will be populated if available.
+
+token-name and library-path can be overridden by passing them on the command-line.
+
+The NSS module name will be the basepath of the library minus .so*.
+
+#### KRA
+
+The KRA transport, audit and storage private keys and certificates are
+generated and stored in the HSM. This is an optional service.
+
+The same token options can be provided to ipa-kra-install.
+
+| Option | Description                                       |
+|:------------------------ | :------------------------------ |
+| --token-name             | NSS name for the token          |
+| --library-path           | Path to PKCS#11 shared library  |
+| --token-password         | Password for the token          |
+
+token-name and library-path can be determined based on the value of ipaCaHSMConfiguration.
+
+### SubCA
+
+SubCA private keys and certificates should be generated and stored in the HSM.
+
+### Replica installation
+
```

Comment:
Since we can only install a replica with a version higher or equal to the master's version, no need to worry about installing a replica which does not support HSM with a server using a HSM.

Reply:
What I mean is you can't install a replica without storing the keys/certs on an HSM if you are already using an HSM in other servers.

Reply:
I added a bit of text under the How to Use section about the basic rules of HSMs.